signalfd.c

```diff
@@ -0,0 +1,170 @@
+ * Written by Dmitry Antipov <dantipov@cloudlinux.com> 2022
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. The name of the author may not be used to endorse or promote products
+ *    derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "event2/event-config.h"
+
+#include <unistd.h>
+#include <sys/signalfd.h>
+
+#include "event2/event.h"
+#include "event-internal.h"
+#include "evmap-internal.h"
+#include "evsignal-internal.h"
+#include "evthread-internal.h"
+
+static int sigfd_add(struct event_base *, evutil_socket_t, short, short, void *);
+static int sigfd_del(struct event_base *, evutil_socket_t, short, short, void *);
+
+static const struct eventop sigfdops = {
+	"signalfd_signal",
+	NULL,
+	sigfd_add,
+	sigfd_del,
+	NULL,
+	NULL,
+	0, 0, 0
+};
+
+static void
+sigfd_cb(evutil_socket_t fd, short what, void *arg)
+{
+	struct signalfd_siginfo fdsi;
+	struct event_base *base = arg;
+	ssize_t ret = read(fd, &fdsi, sizeof(fdsi));
+
+	EVUTIL_ASSERT(ret == sizeof(fdsi));
+	EVUTIL_ASSERT(fdsi.ssi_signo > 0 && fdsi.ssi_signo < NSIG);
+	EVUTIL_ASSERT(base->sig.ev_sigevent[fdsi.ssi_signo] != NULL);
+
+	EVBASE_ACQUIRE_LOCK(base, th_base_lock);
+	evmap_signal_active_(base, fdsi.ssi_signo, 1);
+	EVBASE_RELEASE_LOCK(base, th_base_lock);
+}
+
+static void
+sigfd_free_sigevent(struct event_base *base, int signo)
+{
+	int ret;
+	struct event* sigev = base->sig.ev_sigevent[signo];
+
+	EVUTIL_ASSERT(sigev != NULL);
+	/* Not sure about blocking here. */
+	event_del_nolock_(sigev, EVENT_DEL_AUTOBLOCK);
+	ret = close(sigev->ev_fd);
+	EVUTIL_ASSERT(ret == 0);
+	mm_free(sigev);
+	base->sig.ev_sigevent[signo] = NULL;
+}
+
+static int
+sigfd_add(struct event_base *base, int signo, short old, short events, void *p)
```

Comment:
This is different to `signal`, it does not allow to catch `SIGSEGV` and `SIGFPE`, so maybe we should provide a way to configure this, I'm fine to use it by default, but a way to use plain `signal` will be good to keep

Reply:
I would add per-base `EVENT_BASE_FLAG_DISALLOW_SIGNALFD` flag and check for `ENT_DISALLOW_SIGNALFD` environment variable to unconditionally disable `signalfd`-based signal handling backend.